### ISSUE TITLE
Core ML format

### ISSUE THREAD

**Issue by ivptr** [NathanUA/U-2-Net#118]
Is there an Apple's Core ML format of U^2-Net model?

**Comment by umerzia-7001** [comment]
I tried converting UNet model to coreml format but there is some issue . It is displaying black image in IoS app . When I run prediction on the converted model it gives me no input layer in neural network error. Any solution or coreml model ?

**Comment by BradF-99** [comment]
Would like an update on this as well, if possible. I've also tried converting the model to CoreML 5 using coremltools, but it appears to produce random results:

<img width="1074" alt="Screen Shot 2020-12-13 at 10 47 58 pm" src="https://user-images.githubusercontent.com/37586752/102012403-cfc2be80-3d95-11eb-9743-17a909b027d2.png">

I'm not sure if the issue is in the conversion, or with CoreML itself - 
[I've attached the Jupyter Notebook I created in order to attempt conversion of the model. ](https://github.com/NathanUA/U-2-Net/files/5684595/U2NetCoreMLNotebook.zip)

**Comment by umerzia-7001** [comment]
Hi Brad
Could you be kind enough to provide me the coreml mode you got for the app.
Thnks in advance

**Comment by BradF-99** [comment]
Hi Muhammad,

I've attached the converted U^2net lite model from the Jupyter Notebook. CoreML Tools detected all 7 outputs and flattened them, marking them as MLMultiArray outputs - I've set up extra layers to convert all 7 outputs in to grayscale 320x320 images. The out_p0 output layer is equivalent to d1 (pre-normalisation) from the original model. If needed I can remove the extra layers and provide a model that outputs the default MLMultiArray (which appear to be 1x1x320x320 shape).
[u2netpconv.mlmodel.zip](https://github.com/NathanUA/U-2-Net/files/5684709/u2netpconv.mlmodel.zip)

**Comment by umerzia-7001** [comment]
I need output dimension of MultiArray(Int32 513 x 513) type . Could you do
that ? Input also 513 x 513 dimention

**Comment by umerzia-7001** [comment]
These dimensions and types

**Comment by BradF-99** [comment]
Unfortunately I can't change the input and output sizes, so you may have to do upsampling on the output (I believe the U^2net test Python script uses bilinear upsampling). However, this model accepts a colour image input of 513x513, and supplies outputs as 320x320 MLMultiArray - the "out_a0" output is equivalent to d1. I haven't been able to test it unfortunately, and I still haven't been able to get U^2net full or lite working correctly in CoreML, so it's likely that this model won't work correctly either.

[u2netp513.mlmodel.zip](https://github.com/NathanUA/U-2-Net/files/5685764/u2netp513.mlmodel.zip)

**Comment by MahmoudSalem92** [comment]
Hi Brad, 
Thank you very much for sharing the model. could you share the updated script.
Regards

**Comment by ldenoue** [comment]
@BradF-99 and @MahmoudSalem92 when you run the CoreML model on an iOS device, how fast does it predict? Is it real-time or several seconds per image?

**Comment by BradF-99** [comment]
> 
> 
> @BradF-99 and @MahmoudSalem92 when you run the CoreML model on an iOS device, how fast does it predict? Is it real-time or several seconds per image?

In my testing, I don't think my CoreML converted model would be able to be predict in real-time, it was around a second or two per image. However, I'm using an iPhone X, and I've heard that the Neural Engine in the A11 is not as fast as the newer A-series iterations, and I am not using Obj-C or Swift in my app so it's likely that there are some inefficiencies there. 

It still doesn't get the result right either - maybe it needs some sort of pre-processing on the image. I've just returned from a holiday so I'll look in to it when I have time.

**Comment by MahmoudSalem92** [comment]
@BradF-99 Thank  you very much for your response. Can you upload the code of the generated model?  Does it developed based on python?

**Comment by BradF-99** [comment]
Hi @MahmoudSalem92, the model that was generated is here - [this one converts the output to an image](https://github.com/NathanUA/U-2-Net/files/5684709/u2netpconv.mlmodel.zip), and [this one leaves it as an MLMultiArray](https://github.com/NathanUA/U-2-Net/files/5685764/u2netp513.mlmodel.zip). As for the conversion I used a [Jupyter Notebook](https://github.com/NathanUA/U-2-Net/files/5684595/U2NetCoreMLNotebook.zip).

**Comment by MahmoudSalem92** [comment]
Hi @BradF-99 , thank you very much for your response. the code you shared is generate random results. Do you have the code of this model u2netpconv.mlmodel

**Comment by BradF-99** [comment]
As I stated above I wasn't successful in conversion, clearly it's missing some image processing somewhere. Maybe I've missed something, feel free to look through the Jupyter Notebook I attached.

**Comment by jb-apps** [comment]
Hi @BradF-99 I am getting exactly the same result as you, I think we have to do some preprocessing to the inputs but I haven't been able to figure it out. Have you been able to finding out? Thanks!

**Comment by BradF-99** [comment]
Hi @jb-apps, I did notice that the original Pytorch model requires images in a "BRG" format as opposed to BGR or RGB - maybe this is the pre-processing my CoreML model is missing? Let me know if you have any success, I haven't looked in to this for a while.

**Comment by jb-apps** [comment]
Thanks for the quick response @BradF-99, according to [this](https://github.com/xuebinqin/U-2-Net/issues/176) the inputs should be RGB! We may need to do some pre-processing. I am still new with coreml, will keep investigating.

**Comment by jb-apps** [comment]
Hi @BradF-99 I figured it out what was the problem. We just have to scale the input to be in the range of 0 to 1.
So in your notebook, this would be the changes:
`
traced_model = torch.jit.trace(net, example_input)
_inputs = ct.ImageType(name= "input_1", shape= example_input.shape, scale= 1/255.0)
model = ct.convert(traced_model, inputs=[_inputs])
`
Thank you!

**Comment by TakahiroYamamoto** [comment]
Hi , @BradF-99 ,
Thank you very much for your Jupiter notebook.It became very helpful.
In my case , It was better if I fixed it as follows.Just add scale param.

- before
```
model = ct.convert(traced_model, inputs=[ct.ImageType(name="input_1", shape=example_input.shape)])
```

- fixed
```
model = ct.convert(traced_model, inputs=[ct.ImageType(name="input_1", shape=example_input.shape,scale=1/255.0)])
```

<img width="687" alt="cat" src="https://user-images.githubusercontent.com/6883545/116823459-90b7cd00-abbf-11eb-8118-ead28fa87489.png">

**Comment by iTarek** [comment]
> Hi , @BradF-99 ,
> Thank you very much for your Jupiter notebook.It became very helpful.
> In my case , It was better if I fixed it as follows.Just add scale param.
> 
> * before
> 
> ```
> model = ct.convert(traced_model, inputs=[ct.ImageType(name="input_1", shape=example_input.shape)])
> ```
> 
> * fixed
> 
> ```
> model = ct.convert(traced_model, inputs=[ct.ImageType(name="input_1", shape=example_input.shape,scale=1/255.0)])
> ```
> 
> <img alt="cat" width="687" src="https://user-images.githubusercontent.com/6883545/116823459-90b7cd00-abbf-11eb-8118-ead28fa87489.png">

Hi Can you share the mlmodel?

**Comment by Djeevs** [comment]
> Hi , @BradF-99 ,
> Thank you very much for your Jupiter notebook.It became very helpful.
> In my case , It was better if I fixed it as follows.Just add scale param.
> 
> * before
> 
> ```
> model = ct.convert(traced_model, inputs=[ct.ImageType(name="input_1", shape=example_input.shape)])
> ```
> 
> * fixed
> 
> ```
> model = ct.convert(traced_model, inputs=[ct.ImageType(name="input_1", shape=example_input.shape,scale=1/255.0)])
> ```
> 
> <img alt="cat" width="687" src="https://user-images.githubusercontent.com/6883545/116823459-90b7cd00-abbf-11eb-8118-ead28fa87489.png">

Can you please share the full script you used for conversion?
I looked everywhere and couldn't find any fully working solution.
Also do you guys use Big Sur? I have some problems with Catalina, looks like you need CoreML 5 to change nodes.

**Comment by Anandh-iOS** [comment]
> Hi @MahmoudSalem92, the model that was generated is here - [this one converts the output to an image](https://github.com/NathanUA/U-2-Net/files/5684709/u2netpconv.mlmodel.zip), and [this one leaves it as an MLMultiArray](https://github.com/NathanUA/U-2-Net/files/5685764/u2netp513.mlmodel.zip). As for the conversion I used a [Jupyter Notebook](https://github.com/NathanUA/U-2-Net/files/5684595/U2NetCoreMLNotebook.zip).

Hi Brad,

[Screenshot 2021-06-08 at 6 32 03 PM](https://user-images.githubusercontent.com/12445534/121189642-e8fd8100-c887-11eb-952b-21cbd6c6e981.png)

I am getting this error after conversion using the Jupiter notebook, kindly guide me whats wrong here.

**Comment by whiteio** [comment]
@iTarek @Djeevs @Anandh-iOS This should work.

`import torch
import coremltools as ct
from model import U2NET

net = U2NET(3,1)

net.load_state_dict(torch.load("u2net.pth", map_location=torch.device('cpu')))
net.eval()

example_input = torch.rand(1, 3, 512, 512)
traced_model = torch.jit.trace(net, example_input)

converted_model = ct.convert(
    traced_model, 
    inputs=[ct.ImageType(name="input_1", shape=example_input.shape,scale=1/255.0)]
)

converted_model.save('U2Net.mlmodel')`

**Comment by Anandh-iOS** [comment]
Hi @whiteio,

Thanks for your response, It works and converts to .mlmodel. But I get only a black image on prediction.
//Swift
let resized_image = image?.scaleImage(targetSize: CGSize(width: 512, height: 512))
let buffer = resized_image?.cgImage?.pixelBuffer()
let result = try? u2p!.prediction(input_1: buffer!)         
let image: UIImage = createUIImage(fromFloatArray: try! result!._2169.reshaped(to: [1,512,512]), min: 0, max: 255)!

Kindly share if you find anything wrong here.

Thanks,
Anandh

**Comment by LA-Labs** [comment]
Just created a fully working demo: https://github.com/SmartyToe/Image-segmentation.
Enjoy.

**Comment by Anandh-iOS** [comment]
Hi @LA-Labs,

Thanks for the converted MLModel, it works. If possible please share how you converted the pytorch to .mlmodel.

Thanks.

**Comment by iTarek** [comment]
> Just created a fully working demo: https://github.com/SmartyToe/Image-segmentation.
> Enjoy.

You are the best, Thank you so much for this amazing demo.
Just when compare U^2-Net with DeepLapV3 for removing background, DeepLapV3 is better, is that normal or this version of U^2-Net is not trained version?

**Comment by mgstar1021** [comment]
> Just created a fully working demo: https://github.com/SmartyToe/Image-segmentation.
> Enjoy.

Thanks for your demo! It was helpful for me. However it works for only square image and doesn't support for any ratio. for example, it works for 320*320 but doesn't work for 320 * 300. What is the solution?
